File: src/main/java/guru/xgm/language/angelscript/openapi/options/colors/AngelScriptColorSettingsPage.java
package guru.xgm.language.angelscript.openapi.options.colors;

import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighter;
import com.intellij.openapi.options.colors.AttributesDescriptor;
import com.intellij.openapi.options.colors.ColorDescriptor;
import com.intellij.openapi.options.colors.ColorSettingsPage;
import guru.xgm.language.angelscript.icons.AngelScriptIcons;
import guru.xgm.language.angelscript.openapi.fileTypes.AngelScriptSyntaxHighlighterBase;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.HashMap;
import java.util.Map;

import com.intellij.openapi.options.OptionsBundle;

import static guru.xgm.language.angelscript.openapi.fileTypes.AngelScriptSyntaxHighlighterBase.*;

@SuppressWarnings("DialogTitleCapitalization")
final class AngelScriptColorSettingsPage implements ColorSettingsPage {

    private static final AttributesDescriptor[] DESCRIPTORS = new AttributesDescriptor[]{
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.bad.character"), BAD_CHARACTER_KEY),

            new AttributesDescriptor("Type", TYPE_NAME_KEY),

            new AttributesDescriptor(OptionsBundle.messagePointer("options.language.defaults.line.comment"), LINE_COMMENT_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.language.defaults.block.comment"), BLOCK_COMMENT_KEY),

            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.keyword"), KEYWORD_KEY),

            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.number"), NUMBER_KEY),

            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.parentheses"), PARENTHESES_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.braces"), BRACES_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.brackets"), BRACKETS_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.comma"), COMMA_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.semicolon"), SEMICOLON_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.dot"), DOT_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.operator.sign"), OPERATION_SIGN_KEY),

            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.string"), STRING_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.valid.escape.in.string"), VALID_STRING_ESCAPE_KEY),
            new AttributesDescriptor(OptionsBundle.messagePointer("options.java.attribute.descriptor.invalid.escape.in.string"), INVALID_STRING_ESCAPE_KEY),
    };

    @Override
    public @NotNull Icon getIcon() {
        return AngelScriptIcons.FILE;
    }

    @NotNull
    @Override
    public SyntaxHighlighter getHighlighter() {
        return new AngelScriptSyntaxHighlighterBase();
    }

    @NotNull
    @Override
    public String getDemoText() {
        return """
                <BAD_CHARACTER>$</BAD_CHARACTER>
                // line comment
                /*
                block comment
                */
                <TYPE_NAME>void</TYPE_NAME> d(){
                    if (true){
                        <TYPE_NAME>int</TYPE_NAME> i = 2 * 0xDEADBEAF + '!@#$%^&*' - a.i[2];
                        <TYPE_NAME>string</TYPE_NAME> s = "some<VALID_STRING_ESCAPE>\\n</VALID_STRING_ESCAPE>string<INVALID_STRING_ESCAPE>\\?</INVALID_STRING_ESCAPE>";
                    }
                }
                """;
    }

    @NonNls
    private static final Map<String, TextAttributesKey> ourTags = new HashMap<>() {{
        put("BAD_CHARACTER", BAD_CHARACTER_KEY);
        put("TYPE_NAME", TYPE_NAME_KEY);
        put("VALID_STRING_ESCAPE", VALID_STRING_ESCAPE_KEY);
        put("INVALID_STRING_ESCAPE", INVALID_STRING_ESCAPE_KEY);
    }};

    @Override
    public @NotNull Map<String, TextAttributesKey> getAdditionalHighlightingTagToDescriptorMap() {
        return ourTags;
    }

    @Override
    public AttributesDescriptor @NotNull [] getAttributeDescriptors() {
        return DESCRIPTORS;
    }

    @Override
    public ColorDescriptor @NotNull [] getColorDescriptors() {
        return ColorDescriptor.EMPTY_ARRAY;
    }

    @NotNull
    @Override
    public String getDisplayName() {
        return "AngelScript";
    }

}

File: src/main/kotlin/raft/war/language/angelscript/openapi/options/colors/AngelScriptColorSettingsPage.kt
@file:Suppress("DialogTitleCapitalization")

package raft.war.language.angelscript.openapi.options.colors

import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighter
import com.intellij.openapi.options.OptionsBundle
import com.intellij.openapi.options.colors.AttributesDescriptor
import com.intellij.openapi.options.colors.ColorDescriptor
import com.intellij.openapi.options.colors.ColorSettingsPage
import raft.war.language.angelscript.icons.AngelScriptIcons
import raft.war.language.angelscript.lang.AngelScriptLanguage
import raft.war.language.angelscript.openapi.fileTypes.AngelScriptSyntaxHighlighterBase
import javax.swing.Icon

internal class AngelScriptColorSettingsPage : ColorSettingsPage {
    override fun getIcon(): Icon = AngelScriptIcons.FILE

    override fun getHighlighter(): SyntaxHighlighter = AngelScriptSyntaxHighlighterBase()

    override fun getDemoText(): String {
        return """
                <BAD_CHARACTER>${'$'}</BAD_CHARACTER>
                // line comment
                /*
                block comment
                */
                //${'$'}jass native DestroyGroup takes group whichGroup returns nothing
                <TYPE_NAME>void</TYPE_NAME> d(){
                    if (true){
                        <TYPE_NAME>int</TYPE_NAME> i = 2 * 0xDEADBEAF + '!@#${'$'}%^&*' - a.i[2];
                        <TYPE_NAME>string</TYPE_NAME> s = "some<VALID_STRING_ESCAPE>\
                        </VALID_STRING_ESCAPE>string<INVALID_STRING_ESCAPE>\?</INVALID_STRING_ESCAPE>";
                    }
                }
                
                """.trimIndent()
    }

    override fun getAdditionalHighlightingTagToDescriptorMap(): HashMap<String, TextAttributesKey?> = ourTags

    override fun getAttributeDescriptors(): Array<AttributesDescriptor> = DESCRIPTORS

    override fun getColorDescriptors(): Array<ColorDescriptor> = ColorDescriptor.EMPTY_ARRAY

    override fun getDisplayName(): String = AngelScriptLanguage.NAME

}

private val DESCRIPTORS = arrayOf(
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.bad.character"),
        AngelScriptSyntaxHighlighterBase.BAD_CHARACTER_KEY
    ),

    AttributesDescriptor("Injector", AngelScriptSyntaxHighlighterBase.TYPE_NAME_KEY),

    AttributesDescriptor("Type", AngelScriptSyntaxHighlighterBase.TYPE_NAME_KEY),

    AttributesDescriptor(
        OptionsBundle.messagePointer("options.language.defaults.line.comment"),
        AngelScriptSyntaxHighlighterBase.LINE_COMMENT_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.language.defaults.block.comment"),
        AngelScriptSyntaxHighlighterBase.BLOCK_COMMENT_KEY
    ),

    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.keyword"),
        AngelScriptSyntaxHighlighterBase.KEYWORD_KEY
    ),

    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.number"),
        AngelScriptSyntaxHighlighterBase.NUMBER_KEY
    ),

    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.parentheses"),
        AngelScriptSyntaxHighlighterBase.PARENTHESES_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.braces"),
        AngelScriptSyntaxHighlighterBase.BRACES_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.brackets"),
        AngelScriptSyntaxHighlighterBase.BRACKETS_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.comma"),
        AngelScriptSyntaxHighlighterBase.COMMA_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.semicolon"),
        AngelScriptSyntaxHighlighterBase.SEMICOLON_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.dot"),
        AngelScriptSyntaxHighlighterBase.DOT_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.operator.sign"),
        AngelScriptSyntaxHighlighterBase.OPERATION_SIGN_KEY
    ),

    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.string"),
        AngelScriptSyntaxHighlighterBase.STRING_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.valid.escape.in.string"),
        AngelScriptSyntaxHighlighterBase.VALID_STRING_ESCAPE_KEY
    ),
    AttributesDescriptor(
        OptionsBundle.messagePointer("options.java.attribute.descriptor.invalid.escape.in.string"),
        AngelScriptSyntaxHighlighterBase.INVALID_STRING_ESCAPE_KEY
    ),
)
private val ourTags: HashMap<String, TextAttributesKey?> =
    object : HashMap<String, TextAttributesKey?>() {
        init {
            put("BAD_CHARACTER", AngelScriptSyntaxHighlighterBase.BAD_CHARACTER_KEY)
            put("TYPE_NAME", AngelScriptSyntaxHighlighterBase.TYPE_NAME_KEY)
            put("VALID_STRING_ESCAPE", AngelScriptSyntaxHighlighterBase.VALID_STRING_ESCAPE_KEY)
            put("INVALID_STRING_ESCAPE", AngelScriptSyntaxHighlighterBase.INVALID_STRING_ESCAPE_KEY)
        }
    }
